yogstation/code/modules/scripting/Interpreter/Interpreter.dm

```diff
@@ -121,145 +126,91 @@
-							AssignVariable(name, Eval(stmt.value))
-						else
-							var/datum/D = Eval(GetVariable(stmt.object.id_name))
-							if(!D) return
-							D.vars[stmt.var_name.id_name] = Eval(stmt.value)
+					if(istype(S, /node/expression))
+						. = Eval(S, scope)
 					else if(istype(S, /node/statement/VariableDeclaration))
 						//VariableDeclaration nodes are used to forcibly declare a local variable so that one in a higher scope isn't used by default.
 						var/node/statement/VariableDeclaration/dec=S
-						if(!dec.object)
-							AssignVariable(dec.var_name.id_name, null, curScope)
-						else
-							var/datum/D = Eval(GetVariable(dec.object.id_name))
-							if(!D) return
-							D.vars[dec.var_name.id_name] = null
-					else if(istype(S, /node/statement/FunctionCall))
-						RunFunction(S)
+						scope.init_var(dec.var_name.id_name, src, S)
 					else if(istype(S, /node/statement/FunctionDefinition))
-						//do nothing
+						var/node/statement/FunctionDefinition/dec=S
+						scope.init_var(dec.func_name, new /datum/n_function/defined(dec, scope, src), src, S)
 					else if(istype(S, /node/statement/WhileLoop))
-						RunWhile(S)
+						. = RunWhile(S, scope)
+					else if(istype(S, /node/statement/ForLoop))
+						. = RunFor(S, scope)
 					else if(istype(S, /node/statement/IfStatement))
-						RunIf(S)
+						. = RunIf(S, scope)
 					else if(istype(S, /node/statement/ReturnStatement))
-						if(!curFunction)
-							RaiseError(new/runtimeError/UnexpectedReturn())
+						if(!scope.allowed_status & RETURNING)
+							RaiseError(new/runtimeError/UnexpectedReturn(), scope, S)
 							continue
-						status |= RETURNING
-						returnVal=Eval(S:value)
+						scope.status |= RETURNING
+						. = (scope.return_val=Eval(S:value, scope))
 						break
 					else if(istype(S, /node/statement/BreakStatement))
-						status |= BREAKING
+						if(!scope.allowed_status & BREAKING)
+							//RaiseError(new/runtimeError/UnexpectedReturn())
+							continue
+						scope.status |= BREAKING
 						break
 					else if(istype(S, /node/statement/ContinueStatement))
-						status |= CONTINUING
+						if(!scope.allowed_status & CONTINUING)
+							//RaiseError(new/runtimeError/UnexpectedReturn())
+							continue
+						scope.status |= CONTINUING
 						break
 					else
-						RaiseError(new/runtimeError/UnknownInstruction())
-					if(status)
+						RaiseError(new/runtimeError/UnknownInstruction(S), scope, S)
+					if(scope.status)
 						break
 
-			curScope = scopes.Pop()
-
 /*
 	Proc: RunFunction
 	Runs a function block or a proc with the arguments specified in the script.
 */
-		RunFunction(node/statement/FunctionCall/stmt)
-			//Note that anywhere /node/statement/FunctionCall/stmt is used so may /node/expression/FunctionCall
-
-			// If recursion gets too high (max 50 nested functions) throw an error
-			if(cur_recursion >= MAX_RECURSION)
-				AlertAdmins()
-				RaiseError(new/runtimeError/RecursionLimitReached())
-				return 0
```

Comment:
I see you removed this little anti-recursion thing here. Are you just going to let the BYOND runtime handle stack overflows?

Reply:
no the anti-recursion is handled elsewhere.